### ISSUE TITLE
`dagger call` is tedious with too many flags+args

### ISSUE THREAD

**Issue by sipsma** [dagger/dagger#6723]
Somewhat related to https://github.com/dagger/dagger/issues/6112 but a bit more general

Passing a large number of flags and args when using `dagger call` can become way too tedious to type out by hand. One example would be support for filtering dirs as mentioned [here](https://github.com/dagger/dagger/issues/6155#issuecomment-1961485732). But really any time a function accepts more than a few flags and/or the arguments to the flag end up being verbose are a problem.  

This leaves users reaching for wrappers like shell scripts/makefiles/etc. While that's not the end of the world (i.e. we aren't going to try to stop that), it's not great that users have to layer more abstractions on top of dagger to use it in relatively simple situations.

A few ideas have come up to potentially solve this:
1. `defaultArgs` specified in `dagger.json` as described [here](https://github.com/dagger/dagger/issues/6291#issuecomment-1907368992)
   * Basically, module constructors can have default args encoded in `dagger.json` and those args have the same features as the syntax in `dagger call` (e.g. local files/dirs specified by path, etc.)
2. Support for "recording+saving" `dagger call` invocations and then being able to run them by name. As a strawman for the sake of a concrete example:
   * `dagger call --save-as=my-build -m mydep build --dir '/some/path,**/*.go,!testdata' --some-arg false --token env:MY_TOKEN`
   * After the above, you can now run `dagger call my-build`
3. Support for `dagger script` (https://github.com/dagger/dagger/issues/5774), possibly with an addition to support running scripts written to a file rather than inline.

These aren't mutually exclusive either. I think `2.` might be my preferred starting place at the moment for the sake of being both general+relatively straightforward to implement, but all are pretty tempting and viable.

**Comment by shykes** [comment]
@sipsma FYI I am reviving and refreshing #5774. It is my favorite candidate for solving this issue, I will flesh out my argument and try to build momentum. I think you're right that it's related to #5774, maybe we can kill 2 birds with one stone.

I think we should merge this issue and #6112. The conversation is getting pretty fragmented. WDYT? I'm also going to close #5728.

**Comment by sipsma** [comment]
> @sipsma FYI I am reviving and refreshing #5774. It is my favorite candidate for solving this issue, I will flesh out my argument and try to build momentum. I think you're right that it's related to #5774, maybe we can kill 2 birds with one stone.
> 
> I think we should merge this issue and #6112. The conversation is getting pretty fragmented. WDYT? I'm also going to close #5728.

Yes that all SGTM. I'll also mention that my most recent ideas on how to approach https://github.com/dagger/dagger/issues/5993#issuecomment-1965307584 would also end up being a solution to this problem. Basically, you could write code that executes directly on the host, is able to call out to module dependencies, but is also invokable without relying on toolchains being installed on the host (`go`, `python`, etc.).
* Not mutually exclusive with the dagger script idea though, I'd view dagger script as supporting simple use cases and #5993 being better for more advanced use cases that become hard to encode in a single script

**Comment by shykes** [comment]
Also related: #9584

**Comment by nipuna-perera** [comment]
This is the most recent/relevant discussion on this - https://github.com/dagger/dagger/discussions/10401

**Comment by shykes** [comment]
We're getting close to finally solving this: https://github.com/dagger/dagger/pull/11034
